src/server/models/answer.js

```diff
@@ -0,0 +1,72 @@
+import { DataTypes } from 'sequelize';
+import SequelizeSlugify from 'sequelize-slugify';
+
+import db from '~/server/database';
+import {
+  generateRandomString,
+  generateHashSecret,
+} from '~/server/services/crypto';
+
+export const types = ['property', 'artwork'];
+
+const Answer = db.define('answer', {
+  id: {
+    type: DataTypes.INTEGER,
+    allowNull: false,
+    primaryKey: true,
+    autoIncrement: true,
+  },
+  slug: {
+    type: DataTypes.STRING,
+  },
+  clientId: {
+    type: DataTypes.STRING,
+    allowNull: false,
+    unique: true,
+    validate: {
+      isAlphanumeric: true,
+    },
+  },
+  chainId: {
+    type: DataTypes.STRING,
+    allowNull: false,
+    unique: true,
+    validate: {
+      isAlphanumeric: true,
+    },
+  },
+  secret: {
+    type: DataTypes.STRING,
+    allowNull: false,
+    unique: true,
+    validate: {
+      isAlphanumeric: true,
+    },
+  },
+  type: {
+    type: DataTypes.ENUM,
+    values: types,
+    allowNull: false,
+  },
+});
+
+Answer.hasOne(require('~/server/model/artwork'), {
+  foreignKey: 'slug',
+  allowNull: true,
+});
+Answer.hasOne(require('~/server/model/property'), {
+  foreignKey: 'slug',
+  allowNull: true,
+});
+
+SequelizeSlugify.slugifyModel(Answer, {
+  source: ['clientId'],
+});
+
+Answer.addHook('beforeValidate', async (answer, options) => {
+  answer.clientId = generateRandomString(32);
+  answer.chainId = (await generateHashSecret(options.str)).hash;
+  answer.secret = (await generateHashSecret(options.str)).secret;
+});
```

Comment:
Some thoughts on these lines:

1. The `beforeValidate` hook is called on every update and create, which would indicate that it will generate a new `clientId` etc. on *every* update, is this what we need here or is something like `beforeCreate` better?
2. `generateHashSecret` is not deterministic and whill return a different `hash` and `secret` with every time you run it, even though the `str` is the same. Maybe this is better:

```js
const { hash, secret } = await generateHashSecret(options.str);
answer.chainId = hash;
answer.secret = secret;
```

3. Maybe I'm misunderstanding the Sequelize API but where is `str` coming from in the options object?

Reply:
1 & 2 - yes, thanks for catching. Will change. For 3 - I haven't actually tested this, I'm just sketching out the general idea still, but it's coming from this: https://github.com/sequelize/sequelize/issues/5000 and is a way to pass in a human-readable string from either the artwork or property this answer is associated with